Installation backend from source with pip, mxdev, cookiecutter (#1260)
plone/documentation: docs/install/source-step-by-step.md

We install the Plone backend with `pip`, `cookiecutter-zope-instance`, `mxdev` and other fancy helpers.

```{note}
There will be one single cookiecutter template to install both backend and frontend from scratch. You will find the instructions on {ref}`install-source-installation-jump-label`. That chapter is for you if you want to develop and want to jump in with all steps prepared by an overall cookiecutter. The subsequent sections explain the installation of the backend step by step. You will learn the details of the installation included in the future overall cookiecutter.
```

(install-source-installation-steps-label)=

### Installation steps

Create a new directory to hold your project, make it your current directory, then issue the following commands in a shell session.

Create a Python virtual environment in the current directory.

```shell
python3.9 -m venv venv
source venv/bin/activate
```

Update Python package management tools.

```shell
pip install -U pip wheel
```

````{admonition} packages tree
:class: margin toggle
The Plone packages and dependencies are installed in trees.

> % TODO Why in trees?

Reviewer: To address the question: 

Wheels are different from Eggs here. A Wheel contains exact the structure to be copied over in the tree. 
A tree is more efficient at runtime. An import is faster if the import mechanism can traverse a tree. With Eggs we have several hundred directories with own trees in parallel and on import each of those need to be looked at. 

If it comes to editable (`pip install -e path/to/package`) installs, then it falls back to a mechanism similar to the one known with Eggs: it places a file in the root of the tree with metadata and let it point to the actual directory.

Author: that's very interesting for the understanding.
it skip it anyway, as it goes beyond this chapter target of installation instructions.

```
venv/lib/Python3.x/
site-packages/
┌── plone/
│   ├── …
│   ├── app/
│       ├── caching/
│       ├── content/
│       └── …
…   
├── zope/
│   ├── …
```
````

Install Plone 6 with constrained requirements using `pip`.

```shell
pip install Plone -c https://dist.plone.org/release/6.0.0a4/constraints.txt
```

````{admonition} mkwsgiinstance's minimal Zope configuration
:class: margin toggle
`mkwsgiinstance` creates a home with a minimal configuration for a Zope instance.

> No blob storage, etc. is defined so far.

Reviewer: I would add here: "Plone starts with this configuration, but you would need to edit the files manually to get things like blobstorage for files and images and possible addons working properly."

Author: yes. As this is a sidebar, I even reduce to the skeleton tree without any further text. 
Anything else can be sayd in main content.

```
┌── etc/
│   ├── site.zcml
│   ├── zope.conf
│   └── zope.ini
├── inituser
└── var/
    ├── cache/
    ├── log/
    └── REAMDME.txt/
```
````

Create a Zope instance with the given username and password in the current directory.

(install-source-mkwsgiinstance)=

```shell
mkwsgiinstance -u admin:admin -d .
```

````{admonition} All commands so far
:class: margin

```shell
python3.9 -m venv venv
source venv/bin/activate
pip install -U pip wheel
pip install Plone -c https://dist.plone.org/release/6.0.0a4/constraints.txt
mkwsgiinstance -u admin:admin -d .
```
````

Start the Zope instance.

```shell
runwsgi ./etc/zope.ini
```

You can stop the instance later with {kbd}`ctrl-esc`.

> % TODO How to start the Zope instance in background.

Reviewer: This is covered here with `systemd` https://zope.readthedocs.io/en/latest/operation.html#running-zope-as-a-daemon
Currently there is now alternative to the `instance` script generated by buildout. 
(would be cool if someone invests some time to create such w/o buildout dependency)

Author: I added the section 'process manager'. 

m1c (or 2 cent ;-)) pm2 is a must known

If you now open the browser with http://localhost:8080/, you see that you already can create a Plone instance.
Before doing this, we configure our Zope instance for blobs, configure add-ons, etc..
For the configuration, you have two options:
- manual configuration of site.zcml and zope.conf (^[Configuring and Running Zope](https://zope.readthedocs.io/en/latest/operation.html))

> - apply `cookiecutter-zope-instance`

Reviewer: ```suggestion
1. *manual* configuration by editing {file}`site.zcml` and {file}`zope.conf` (^[Configuring and Running Zope](https://zope.readthedocs.io/en/latest/operation.html))
2. *generate* configuration by applying `cookiecutter-zope-instance`
```

(install-source-cookiecutter-zope-instance-label)=

#### Cookiecutter Zope instance

> `Cookiecutter` creates projects from project templates.

Reviewer: This should go to glossary
```suggestion
{term}`Cookiecutter` creates projects from project templates.
```

> {term}`cookiecutter-zope-instance` is such a template that allows to create a complete Zope configuration. Zope configuration means: blob storage, type of file storage (Zope style or relational database), ZEO, you name it.

Reviewer: ```suggestion
{term}`cookiecutter-zope-instance` is such a template that allows to create a complete Zope configuration. 
Zope configuration means: 
- type of storage: direct filestorage/blobs, relational database, ZEO, you name it.
- ports, 
- threads/caches,
- debugging/profiling options, 
- ...
```

Install cookiecutter:

```shell
pip install cookiecutter
```

You could now run `cookiecutter` to create a Zope instance sceleton including configuration with the following command. It prompts you for parameter values.

```shell
cookiecutter https://github.com/plone/cookiecutter-zope-instance
```

(install-source-cookiecutter-zope-instance-presets-label)=

Instead we prepare a file {file}`instance.yaml` with the parameters we want to set. 

> A minimal example is (add options as needed):

Reviewer: ```suggestion
A minimal example with one add-on configured is (add options as needed):
```

```yaml
default_context:
    initial_user_name: 'admin'
    initial_user_password: 'admin'

    load_zcml:
        package_includes: ['collective.easyform']

    db_storage: direct
```

Find more [options of cookiecutter `cookiecutter-zope-instance`](https://github.com/plone/cookiecutter-zope-instance#options).

The file {file}`instance.yaml` allows to set some presets.
Add-ons are listed here, but need to be installed with pip.
The documented installation of add-ons with pip is achieved via a {file}`requirements.txt` file.
An add-on like for example `collective.easyform` is listed in\
`requirements.txt`:

```
collective.easyform
```

Install your requirements:

```shell
pip install -r requirements.txt
```

You have done two things so far: You installed your add-on packages and you have prepared an initializing file to roll out a Zope / Plone project, configured to load your installed add-on packages.

You are now ready to apply your cookiecutter:

```shell
cookiecutter -f --no-input --config-file instance.yaml https://github.com/plone/cookiecutter-zope-instance
```

> ```{list-table} Cookiecutter options

Reviewer: ```suggestion
```{list-table} Cookiecutter options used here
```

:header-rows: 1

* - Option
  - Effect
* - -f, --overwrite-if-exists
  - Overwrite the contents of the output directory if it already exists. Data in var stays untouched.
* - --no-input
  - Do not prompt for parameters and only use cookiecutter.json file content
* - --config-file instance.yaml
  - User configuration file
```

````{admonition} Summed up tasks and commands
:class: margin

Summed up tasks and commands after creating a Zope instance with {ref}`mkwsgiinstance <install-source-mkwsgiinstance>`:

- Create an {ref}`instance.yaml<install-source-cookiecutter-zope-instance-presets-label>` for presets.
- Create a `requirements.txt` for the installation of python packages.
- Install add-ons and generate Zope skeleton with configuration (apply cookiecutter)

  ```shell
  pip install -r requirements.txt
  cookiecutter -f --no-input --config-file instance.yaml https://github.com/plone/cookiecutter-zope-instance
  ```
````

`cookiecutter` creates a new directory {file}`instance` with your Zope and Plone configuration. This directory will also comprise file storage, blob storage, etc..

It's time to start the new Zope instance.

```shell

> runwsgi instance/etc/zope.ini

Reviewer: shall we tell here that `runwsgi -v instance/etc/zope.ini` creates a more verbose output?

Reviewer 2: Yes, we should

Author: replaced by `runwsgi -v instance/etc/zope.ini` with the flag for a verbose output

```

Head over to http://localhost:8080/ and see that Plone is running.
You could now create a Plone instance {ref}`install-source-create-plone-site-label` and enable the add-on `collective.easyform` in {guilabel}`Site Setup` [http://localhost:8080/Plone/prefs_install_products_form](http://localhost:8080/Plone/prefs_install_products_form).

If you want to run Plone with some add-ons, you are ready with the installation of the backend.\
If you decided to go with the Plone Classic UI this is also your frontend.\
If you decided to go with the Plone Volto frontend, then section {ref}`install-source-volto-frontend-label` explains your next steps.

If you want to develop a Plone package, then the subsequent section is for you.

(install-source-checkout-and-pin)=

#### Checkout or version pinning of a Plone package

> If you want to checkout a Plone package (not add-on) for development or just want to override the constraints of Plone, then a first attempt would be to define constraints with a {file}`constraints.txt` to tell pip to install a different version of a Plone package.

Reviewer: I dont understand the `(not addon)`  - I think later on this this is valid for addons as well?

I would say `If you want to checkout a Plone Core package for development`

Reviewer: BUT you can do this for any addon as well.

```
# constraints.txt with unresolvable version conflict
-c https://dist.plone.org/release/6.0.0a4/constraints.txt
plone.api>=2.0.0a3
```

Unfortunatly pip does not allow this way of overwriting constraints. 

`mxdev` is made for assembling Plone constraints with your needs of version pinning or source checkouts.\

> It reads your {file}`constraints.txt`, fetches the constraints of Plone and writes a {file}`constraints-mxdev.txt` which combines the constraints. Comments on which Plone constraint is modified assure the readability.
> % TODO language: 'readability'?

Reviewer: Idea: 
```suggestion
It reads your {file}`requirements.txt` and so your {file}`constraints.txt`, fetches the requirements/constraints of Plone and writes the files {file}`requirements-mxdev.txt` and {file}`constraints-mxdev.txt`. 
Both are containing the combined requirements and constraints, but modified according to the configuration in {file}`mx.ini`. 
The generated files are transparent about where constraints were fetched from and comments are added when a modification was necessary.
```

Reviewer: OTOH this may get removed here and added to the sentence below?

Author: yes, thanks, rephrased it.

mxdev is made for assembling Plone constraints with your needs of version pinning or source checkouts.
It reads your requirements.txt and so your constraints.txt, fetches the requirements/constraints of Plone and writes the files requirements-mxdev.txt and constraints-mxdev.txt. Both are containing the combined requirements and constraints, but modified according to the configuration in mx.ini. The generated files are transparent about where constraints were fetched from and comments are added when a modification was necessary.

In summary, mxdev operates on three files to tell pip which packages to install with which version. A minimal example set of files would look like the following.

mxdev operates on three files to tell pip which packages to install with which version.

{file}`requirements.txt`

```ini
-c constraints.txt

Plone

# List of add-ons that are needed.
collective.easyform
```

{file}`constraints.txt`

```ini
-c https://dist.plone.org/release/6.0.0a4/constraints.txt

# constraints of add-ons
collective.easyform==3.4.5
```

{file}`mx.ini`

```ini
[settings]
# constraints of Plone packages
version-overrides =
    plone.api>=2.0.0a3

[plone.restapi]
url = git@github.com:plone/plone.restapi.git
branch = master
extras = test
```

A run of mxdev reads {file}`requirements.txt`, {file}`constraints.txt` and {file}`mx.ini`, and writes new combined requirements in {file}`requirements-mxdev.txt` and writes new constraints in {file}`constraints-mxdev.txt` according to {file}`mx.ini`:

> - 'version-overrides' in [settings]
> - checkout settings in [packagename] sections

Reviewer: ```suggestion
- 'version-overrides' in section `[settings]`
- checkout settings in the `[packagename]` sections
```

{term}`Cookiecutter` creates projects from project templates.
{term}`cookiecutter-zope-instance` is such a template that allows to create a complete Zope configuration. 
Zope configuration means:

- type of storage: direct filestorage/blobs, relational database, ZEO, you name it.
- ports, 
- threads/caches,
- debugging/profiling options, 
- ...

Install cookiecutter:

```shell
pip install cookiecutter
```

You could now run `cookiecutter` to create a Zope instance sceleton including configuration with the following command. It prompts you for parameter values.

```shell
cookiecutter https://github.com/plone/cookiecutter-zope-instance
```

(install-source-cookiecutter-zope-instance-presets-label)=

Instead we prepare a file {file}`instance.yaml` with the parameters we want to set. 
A minimal example with one add-on configured is (add options as needed):

```yaml
default_context:
    initial_user_name: 'admin'
    initial_user_password: 'admin'

    load_zcml:
        package_includes: ['collective.easyform']

    db_storage: direct
```

Find more [options of cookiecutter `cookiecutter-zope-instance`](https://github.com/plone/cookiecutter-zope-instance#options).

> The file {file}`instance.yaml` allows to set some presets.

Reviewer 2: add "... and in this section we will learn how to use those presets".

Author: rephrased and added something similar.

Add-ons are listed here, but need to be installed with pip.

> The documented installation of add-ons with pip is achieved via a {file}`requirements.txt` file.

Reviewer 2: if add-ons are listed here in instance.yaml but need to be installed with pip using a requirements.txt file, why would we need to list them? An explanation should be added.

Author: yes, thanks. I rephrased it.

> ---

Reviewer 3: Per an earlier review discussion, wherein we changed the title of this file, we should also change the filename to align with it.

Also I don't think "step-by-step" is needed (as an adjective, it is hyphenated, whereas in other usage it might have spaces instead).

Thus let's rename this file to `backend-packages.md`.

Author: The reason why I have chosen the addition to the title "step by step" was, that I wrote two chapters with installation instructions according to the fact, that the recommended way, how to install the backend, is still not fixed. The plan is to have one cookiecutter that does most of what I describe in the step by step chapter. Even when this cookiecutter exists, the "step by step" chapter (naming can be changed, of course) is still valid and of interest for in depth knowledge.

Reviewer 3: This confuses me further. Are these the two chapters you mentioned?

- [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#)
- [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html)

In the screenshot:

![Screen Shot 2022-07-23 at 3 21 33 AM](https://user-images.githubusercontent.com/102112/180601278-5b49959e-b435-446d-a032-d09b58333c50.png)

1. I click [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html).
2. I see that I need to [Install the Plone backend](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#install-source-install-backend-pip-label). Do I click that, or...
3. I see there is a navigation item to [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html#). Do I click this one?

It sounds like your intent is for the new-to-Plone developer to follow the instructions [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#). Then if they want details of that process, they may refer to [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html). Is that what you intend?

As it stands, the naming of these two files and their introductory content makes me pause with uncertainty of how to proceed.

Author: The challenge is to provide installation instructions while one installation tool is still to be written. 
The gap that needs to be filled in the documentation, because the installation tool is not ready, is here:
https://6.dev-docs.plone.org/install/source.html#installation-jump-in-and-enfold-the-beauty
![image](https://user-images.githubusercontent.com/1144737/181010918-e5f6abf8-1386-461d-96b3-ac2bb9c954cc.png)
So I wrote the chapter https://6.dev-docs.plone.org/install/source.html as it would be with an existing cookiecutter for a backend installation (or a backend and frontend installation all in one) and placed the note https://6.dev-docs.plone.org/install/source.html#installation-jump-in-and-enfold-the-beauty for information. And this note also links to the instructions https://6.dev-docs.plone.org/install/source-step-by-step.html, how to install the backend in the mean time until this future tool exists. 

Apart from the current situation of installation tools, the chapter https://6.dev-docs.plone.org/install/source-step-by-step.html could stay in place for detailed information about what this future tool does under the hood.

I hope this makes the construction a little bit clearer. 

 

> This confuses me further. Are these the two chapters you mentioned?
> 
>     * [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#)
> 
>     * [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html)
> 
> 
> In the screenshot:
> 
> ![Screen Shot 2022-07-23 at 3 21 33 AM](https://user-images.githubusercontent.com/102112/180601278-5b49959e-b435-446d-a032-d09b58333c50.png)
> 

This screenshot shows on the right side the anchor navigation of this main chapter. 
I now read you comment and the docu multiple times and I think the supplementary nature of https://6.dev-docs.plone.org/install/source-step-by-step.html would be more obvious if https://6.dev-docs.plone.org/install/source-step-by-step.html would become a subchapter of the main https://6.dev-docs.plone.org/install/source.html What do you think?

>     1. I click [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html).
> 
>     2. I see that I need to [Install the Plone backend](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#install-source-install-backend-pip-label). Do I click that, or...
> 
>     3. I see there is a navigation item to [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html#). Do I click this one?
> 
> 
> It sounds like your intent is for the new-to-Plone developer to follow the instructions [Install Plone from its Packages](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-from-packages.html#). Then if they want details of that process, they may refer to [Install Plone backend from its Packages – Step-by-step](https://deploy-preview-1274--6-dev-docs-plone-org.netlify.app/install/installation-backend-from-packages-step-by-step.html). Is that what you intend?
> 
> As it stands, the naming of these two files and their introductory content makes me pause with uncertainty of how to proceed.

html_meta:
  "description": "Install Plone 6 backend from its packages for the one who wants to look under the hood"
  "property=og:description": "Install Plone 6 backend from its packages for the one who wants to look under the hood"
  "property=og:title": "Install Plone backend from its Packages – Step by Step"
  "keywords": "Plone, Plone 6, install, backend, pip, mxdev, mxmake, cookiecutter, packages, source, Zope, buildout"
---

> (install-source-stepbystep-start-label)=

Reviewer 3: ```suggestion
(install-backend-packages-label)=
```

> # Install Plone backend from its Packages – Step by Step

Reviewer 3: ```suggestion
# Install Plone backend from its packages
```

> (install-source-stepbystep-backend-start-label)=

Reviewer 3: ```suggestion
```

> ## Backend

Reviewer 3: ```suggestion
```

Reviewer 3: The following content makes for a good introduction instead of a sub-section. Let's promote it a heading level.

For system requirements and pre-requisites for the installation see {ref}`install-source-system-requirements-label`.

> We install the Plone backend with `pip`, `cookiecutter-zope-instance`, `mxdev` and other fancy helpers.

Reviewer 3: ```suggestion
We install the Plone backend with `pip`, `cookiecutter-zope-instance`, `mxdev`, and other developer tools.
```

```{note}

> There will be one single cookiecutter template to install both backend and frontend from its packages. You will find the instructions on {ref}`install-source-installation-jump-label`. That chapter is for you if you want to develop and want to jump in with all steps prepared by an overall cookiecutter. The subsequent sections explain the installation of the backend step by step. You will learn the details of the installation included in the future overall cookiecutter.

Reviewer 3: Break into one sentence per line.

Reviewer 3: This note is somewhat confusing. Has the referenced cookiecutter been released and completed? If not, then we should explicitly state that it is under active development, alpha, beta, or whatever is its status now.

Once I know its status, I can help rewrite this to be more explicit.

```

> (install-source-installation-steps-label)=

Reviewer 3: ```suggestion
(install-backend-packages-installation-steps-label)=
```

> ### Installation steps

Reviewer 3: ```suggestion
## Installation steps
```

> Create a new directory to hold your project, make it your current directory, then issue the following commands in a shell session.

Reviewer 3: ```suggestion
Create a new directory to hold your project, and make it your current directory.
Next issue the following commands in a shell session to create a Python virtual environment in the current directory.
```

> Create a Python virtual environment in the current directory.

Reviewer 3: ```suggestion
```

```shell

> python3.9 -m venv venv

Reviewer 3: Future proof our python command.

```suggestion
python3 -m venv venv
```

source venv/bin/activate
```

Update Python package management tools.

```shell
pip install -U pip wheel
```

> ````{admonition} packages tree

Reviewer 3: This section appears too early in the page. At this moment, only the virtual environment has been created and pip and wheel updated. It should be moved after the command that creates it.

Reviewer 3: ```suggestion
````{note} The Plone packages and dependencies are installed in a directory tree.
```

Author: yes, the order is confusing. It's just this way to have the admonition on the rendered page height with the line "Install Plone 6 with constrained requirements using `pip`."

> :class: margin toggle

Reviewer 3: ```suggestion
:class: margin
```

> The Plone packages and dependencies are installed in trees.

Reviewer 3: ```suggestion
```

```
venv/lib/Python3.x/
site-packages/

> ┌── plone/

Reviewer 3: This tree diagram is confusing.

````suggestion
```
venv/lib/python3.x/site-packages
│
…
├── plone
│   ├── app
│   │   ├── caching
│   │   ├── content
…
├── zope
│   ├── annotation
│   ├── app
…
```
````

│   ├── …
│   ├── app/
│       ├── caching/
│       ├── content/
│       └── …
…   
├── zope/
│   ├── …
```
````

Install Plone 6 with constrained requirements using `pip`.

```shell
pip install Plone -c https://dist.plone.org/release/6.0.0a6/constraints.txt
```

> ````{admonition} mkwsgiinstance's minimal Zope configuration

Reviewer 3: This should appear after its command.

```suggestion
````{note} `mkwsgiinstance` creates a home with a minimal configuration for a Zope instance.
```

> :class: margin toggle

Reviewer 3: ```suggestion
:class: margin
```

> `mkwsgiinstance` creates a home with a minimal configuration for a Zope instance.

Reviewer 3: ```suggestion
```

```
┌── etc/
│   ├── site.zcml
│   ├── zope.conf
│   └── zope.ini
├── inituser
└── var/
    ├── cache/
    ├── log/
    └── REAMDME.txt/
```
````

> Create a Zope instance with the given username and password in the current directory.

Reviewer 3: ```suggestion
(install-backend-packages-mkwsgiinstance-label)=

Create a Zope instance with the given username and password in the current directory.
```

Reviewer 3: Move the label up above the command's explanatory text to provide context.

> (install-source-mkwsgiinstance)=

Reviewer 3: ```suggestion
```

```shell
mkwsgiinstance -u admin:admin -d .
```

> ````{admonition} All commands so far

Reviewer 3: This admonition is redundant. Please remove.

Author: It's a summary.

:class: margin

```shell
python3.9 -m venv venv
source venv/bin/activate
pip install -U pip wheel
pip install Plone -c https://dist.plone.org/release/6.0.0a6/constraints.txt
mkwsgiinstance -u admin:admin -d .
```
````

Start the Zope instance.

```shell
runwsgi ./etc/zope.ini
```

> You can stop the instance later with {kbd}`ctrl-esc`.

Reviewer 3: ```suggestion
It will take a few seconds to start the Zope instance.
You can stop the instance later with {kbd}`ctrl-esc`.
```

> If you now open the browser with http://localhost:8080/, you see that you already can create a Plone instance.

Reviewer 3: ```suggestion
If you open the browser and visit http://localhost:8080/, you are presented with several choices.
```

Note that the text is not correct.

![Screen Shot 2022-06-29 at 5 20 08 AM](https://user-images.githubusercontent.com/102112/176435249-931da994-7dfc-42fe-a3db-0cea879570c7.png)

> For an introduction to Plone, success stories, demos, providers, visit [plone.com](http://plone.com/). For documentation, add-ons, support, community, visit [plone.org](http://plone.org/).

plone.com redirects to plone.org. Where should I report that issue?

Reviewer 3: I'm going to stop here for tonight.

The next few lines confuse me and make me slam on the brakes. I open a web browser, and what exactly am I supposed to do? Click a button? Edit files? Use a cookiecutter?

Perhaps I should not visit the URL in the first place, and instead focus on configuration first?

Reviewer: > plone.com redirects to plone.org. Where should I report that issue?

https://github.com/plone/Products.CMFPlone/issues - thanks!

Author: to reduce the confusion, I shortened the two sentences to 
_Before creating a Plone site on http://localhost:8080/, we configure our Zope instance for blobs, configure add-ons, etc.._

---
html_meta:
  "description": "Install Plone 6 from source for the one who wants to look under the hood"
  "property=og:description": "Install Plone 6 from source for the one who wants to look under the hood"
  "property=og:title": "Install Plone 6 from source step by step"
  "keywords": "Plone 6, install, buildout, pip, mxdev, mxmake, cookiecutter, source, Zope"
---

(install-source-stepbystep-start-label)=

# Install Plone Backend from Source – Step by Step

> For system requirements and Pre-requisites for the installation see {ref}`install-source-system-requirements-label`.

Reviewer 3: ```suggestion
For system requirements and pre-requisites for the installation, see {ref}`install-source-system-requirements-label`.
```

> We install the Plone backend with `pip`, `cookiecutter-zope-instance`, `mxstack` and other fancy helpers.

Reviewer 3: ```suggestion
We install the Plone backend with `pip`, `cookiecutter-zope-instance`, `mxstack`, and other fancy helpers.
```

(install-source-installation-steps-label)=

> ### Installation steps

Reviewer 3: ```suggestion
## Installation steps
```

> ```{admonition} Jump in!

Reviewer 3: This totally throws me off. I read a heading for one step, then I am immediately given an option to jump into another step. I think we need an introduction to the various workflows before this section starts.
